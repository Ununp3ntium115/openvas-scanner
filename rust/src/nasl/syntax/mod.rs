// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

#![doc = include_str!("README.md")]
mod cursor;
mod error;
mod grouping_extension;
mod keyword_extension;
mod lexer;
mod loader;
mod naslvalue;
mod operation;
mod prefix_extension;
mod statement;
mod token;
mod tokenizer;
mod variable_extension;

pub use crate::storage::item::ACT;
pub use error::{ErrorKind, SyntaxError};
pub use lexer::Lexer;
pub use loader::*;
pub use naslvalue::*;
pub use statement::*;
pub use token::Keyword;
pub use token::Token;
pub use token::TokenKind;
pub use tokenizer::Tokenizer;

/// Parses given code and returns found Statements and Errors
///
/// # Examples
/// Basic usage:
///
/// ```
/// use scannerlib::nasl::syntax::{Statement, SyntaxError, parse};
/// let statements =
///     parse("a = 23;b = 1;").collect::<Vec<Result<Statement, SyntaxError>>>();
/// ````
pub fn parse(code: &str) -> impl Iterator<Item = Result<Statement, SyntaxError>> + '_ {
    // TODO Do not unwrap here, handle errors properly.
    let tokens = Tokenizer::tokenize(code).unwrap();
    Lexer::new(tokens)
}

#[cfg(test)]
mod tests {
    use crate::nasl::syntax::{
        cursor::Cursor,
        token::{Keyword, Token, TokenKind},
        Tokenizer,
    };

    #[test]
    fn use_cursor() {
        let mut cursor = Cursor::new("  \n\tdisplay(12);");
        cursor.skip_while(|c| c.is_whitespace());
        assert_eq!(cursor.advance(), Some('d'));
    }

    #[test]
    fn use_tokenizer() {
        let all_tokens = Tokenizer::tokenize("local_var hello = 'World!';").unwrap();
        assert_eq!(
            all_tokens,
            vec![
                Token {
                    kind: TokenKind::Identifier(Keyword::LocalVar),
                    line_column: (1, 1),
                    position: (0, 9)
                },
                Token {
                    kind: TokenKind::Identifier(Keyword::Undefined("hello".to_owned())),
                    line_column: (1, 11),
                    position: (10, 15)
                },
                Token {
                    kind: TokenKind::Equal,
                    line_column: (1, 17),
                    position: (16, 17)
                },
                Token {
                    kind: TokenKind::Data("World!".as_bytes().to_vec()),
                    line_column: (1, 19),
                    position: (18, 26)
                },
                Token {
                    kind: TokenKind::Semicolon,
                    line_column: (1, 27),
                    position: (26, 27)
                },
            ]
        );
    }

    #[test]
    fn use_parser() {
        let code = "a = 23;b = 1;";
        let expected = ["a = 23;", "b = 1;"];
        for (i, s) in super::parse(code).enumerate() {
            let stmt = s.unwrap();
            //assert!(matches!(stmt.kind(), Assign(..)));
            assert_eq!(&code[stmt.range()], expected[i]);
        }
    }
}
