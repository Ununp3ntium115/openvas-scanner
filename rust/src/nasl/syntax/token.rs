// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

//! This module defines the TokenTypes as well as Token and extends Cursor with advance_token
use std::{fmt::Display, net::Ipv4Addr};

use crate::{
    nasl::{error::Span, interpreter::InterpretError},
    storage::items::nvt::ACT,
};

use super::CharIndex;

/// A reserved NASL keyword.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Keyword {
    /// function declaration
    Function,
    /// _FCT_ANON_ARGS
    FCTAnonArgs,
    /// for
    For,
    /// foreach
    ForEach,
    /// if
    If,
    /// else
    Else,
    /// while
    While,
    /// repeat
    Repeat,
    /// until
    Until,
    /// local_var
    LocalVar,
    /// global_var
    GlobalVar,
    /// return
    Return,
    /// continue
    Continue,
    /// break
    Break,
    /// include
    Include,
    /// Scanning phases; can be set by category in the description block
    ACT(ACT),
    /// exit
    Exit,
}

macro_rules! make_keyword_matcher {
    ($($matcher:ident, $define:expr),+) => {
        impl Keyword {
            /// Creates a new keyword based on a string identifier
            pub fn new(keyword: &str) -> Option<Self> {
                match keyword {
                    $(
                    stringify!($matcher) => Some($define),
                    )*
                    _ => None,
                }
            }
        }

        impl std::fmt::Display for Keyword {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                // $define is not a pattern but expr, so in the
                // name of simplicity we use a chained if here
                $(
                    if *self == $define {
                        return write!(f, stringify!($matcher));
                    }
                )*
                unreachable!()
            }
        }
    }
}

make_keyword_matcher! {
    function, Keyword::Function,
    _FCT_ANON_ARGS, Keyword::FCTAnonArgs,
    for, Keyword::For,
    foreach, Keyword::ForEach,
    if, Keyword::If,
    else, Keyword::Else,
    while, Keyword::While,
    repeat, Keyword::Repeat,
    until, Keyword::Until,
    local_var, Keyword::LocalVar,
    global_var, Keyword::GlobalVar,
    return, Keyword::Return,
    include, Keyword::Include,
    exit, Keyword::Exit,
    ACT_ATTACK, Keyword::ACT(ACT::Attack),
    ACT_DENIAL, Keyword::ACT(ACT::Denial),
    ACT_DESTRUCTIVE_ATTACK, Keyword::ACT(ACT::DestructiveAttack),
    ACT_END, Keyword::ACT(ACT::End),
    ACT_FLOOD, Keyword::ACT(ACT::Flood),
    ACT_GATHER_INFO, Keyword::ACT(ACT::GatherInfo),
    ACT_INIT, Keyword::ACT(ACT::Init),
    ACT_KILL_HOST, Keyword::ACT(ACT::KillHost),
    ACT_MIXED_ATTACK, Keyword::ACT(ACT::MixedAttack),
    ACT_SCANNER, Keyword::ACT(ACT::Scanner),
    ACT_SETTINGS, Keyword::ACT(ACT::Settings),
    continue, Keyword::Continue,
    break, Keyword::Break
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Ident(pub String);

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Literal {
    /// A String (")
    ///
    /// Strings can be over multiple lines and are not escapable (`a = "a\";` is valid).
    /// A string type will be cast to utf8 string.
    String(String),
    /// Data is defined Quotable (')
    ///
    /// Data can be over multiple lines and are escaped (`a = "a\";` is valid).
    /// Unlike string the data types are stored in bytes.
    Data(Vec<u8>),
    /// A Number can be either binary (0b), octal (0), base10 (1-9) or hex (0x)
    Number(i64),
    /// An IP address. We currently just support 127.0.0.1 notation
    IPv4Address(Ipv4Addr),
    /// A boolean.
    Boolean(bool),
    /// Null
    Null,
}

impl Literal {
    pub fn from_keyword(lookup: &str) -> Option<Self> {
        match lookup {
            "NULL" => Some(Self::Null),
            "FALSE" => Some(Self::Boolean(false)),
            "TRUE" => Some(Self::Boolean(true)),
            _ => None,
        }
    }
}

/// Is used to identify a Token
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum TokenKind {
    /// `(`
    LeftParen,
    /// `)`
    RightParen,
    /// `[`
    LeftBracket,
    /// `]`
    RightBracket,
    /// `{`
    LeftBrace,
    /// `}`
    RightBrace,
    /// `,`
    Comma,
    /// `.`
    Dot,
    /// `%`
    Percent,
    /// `%=`
    PercentEqual,
    /// `;`
    Semicolon,
    /// `:`
    DoublePoint,
    /// `~`
    Tilde,
    /// `^`
    Caret,
    /// `&`
    Ampersand,
    /// `&&`
    AmpersandAmpersand,
    /// `|`
    Pipe,
    /// `||`
    PipePipe,
    /// `!`
    Bang,
    /// `!=`
    BangEqual,
    /// `!~`
    BangTilde,
    /// `=`
    Equal,
    /// `==`
    EqualEqual,
    /// `=~`
    EqualTilde,
    /// `>`
    Greater,
    /// `>>`
    GreaterGreater,
    /// `>=`
    GreaterEqual,
    /// `><`
    GreaterLess,
    /// `<`
    Less,
    /// `<<`
    LessLess,
    /// `<=`
    LessEqual,
    /// `-`
    Minus,
    /// `--`
    MinusMinus,
    /// `-=`
    MinusEqual,
    /// `+`
    Plus,
    /// `+=`
    PlusEqual,
    /// `++`
    PlusPlus,
    /// `/`
    Slash,
    /// `/=`
    SlashEqual,
    /// `*`
    Star,
    /// `**`
    StarStar,
    /// `*=`
    StarEqual,
    /// `>>>`
    GreaterGreaterGreater,
    /// `>>=`
    GreaterGreaterEqual,
    /// `<<=`
    LessLessEqual,
    /// `>!<`
    GreaterBangLess,
    /// `>>>=`
    GreaterGreaterGreaterEqual,
    /// `x` is a special functionality to redo a function call n times.E.g. `send_packet( udp, pcap_active:FALSE ) x 200;`
    X,
    /// A literal, such as a number, string, data or IP address
    Literal(Literal),
    /// Special keywords reserved within NASL.
    Keyword(Keyword),
    /// An identifier for a variable or function.
    Ident(Ident),
    /// End of file
    Eof,
}

impl Display for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenKind::LeftParen => write!(f, "("),
            TokenKind::RightParen => write!(f, ")"),
            TokenKind::LeftBracket => write!(f, "["),
            TokenKind::RightBracket => write!(f, "]"),
            TokenKind::LeftBrace => write!(f, "{{"),
            TokenKind::RightBrace => write!(f, "}}"),
            TokenKind::Comma => write!(f, ","),
            TokenKind::Dot => write!(f, "."),
            TokenKind::Percent => write!(f, "%"),
            TokenKind::PercentEqual => write!(f, "%="),
            TokenKind::Semicolon => write!(f, ";"),
            TokenKind::DoublePoint => write!(f, ":"),
            TokenKind::Tilde => write!(f, "~"),
            TokenKind::Caret => write!(f, "^"),
            TokenKind::Ampersand => write!(f, "&"),
            TokenKind::AmpersandAmpersand => write!(f, "&&"),
            TokenKind::Pipe => write!(f, "|"),
            TokenKind::PipePipe => write!(f, "||"),
            TokenKind::Bang => write!(f, "!"),
            TokenKind::BangEqual => write!(f, "!="),
            TokenKind::BangTilde => write!(f, "!~"),
            TokenKind::Equal => write!(f, "="),
            TokenKind::EqualEqual => write!(f, "=="),
            TokenKind::EqualTilde => write!(f, "=~"),
            TokenKind::Greater => write!(f, ">"),
            TokenKind::GreaterGreater => write!(f, ">>"),
            TokenKind::GreaterEqual => write!(f, ">="),
            TokenKind::GreaterLess => write!(f, "><"),
            TokenKind::Less => write!(f, "<"),
            TokenKind::LessLess => write!(f, "<<"),
            TokenKind::LessEqual => write!(f, "<="),
            TokenKind::Minus => write!(f, "-"),
            TokenKind::MinusMinus => write!(f, "--"),
            TokenKind::MinusEqual => write!(f, "-="),
            TokenKind::Plus => write!(f, "+"),
            TokenKind::PlusEqual => write!(f, "+="),
            TokenKind::PlusPlus => write!(f, "++"),
            TokenKind::Slash => write!(f, "/"),
            TokenKind::SlashEqual => write!(f, "/="),
            TokenKind::Star => write!(f, "*"),
            TokenKind::StarStar => write!(f, "**"),
            TokenKind::StarEqual => write!(f, "*="),
            TokenKind::GreaterGreaterGreater => write!(f, ">>>"),
            TokenKind::GreaterGreaterEqual => write!(f, ">>="),
            TokenKind::LessLessEqual => write!(f, "<<="),
            TokenKind::GreaterBangLess => write!(f, ">!<"),
            TokenKind::GreaterGreaterGreaterEqual => write!(f, ">>>="),
            TokenKind::X => write!(f, "X"),
            TokenKind::Keyword(kw) => write!(f, "{}", kw),
            TokenKind::Ident(ident) => write!(f, "{}", ident.0),
            TokenKind::Literal(Literal::Number(num)) => write!(f, "{num}"),
            TokenKind::Literal(Literal::String(s)) => write!(f, "\"{s}\""),
            TokenKind::Literal(Literal::Data(data)) => write!(f, "{data:?}"),
            TokenKind::Literal(Literal::IPv4Address(ip)) => write!(f, "{ip}"),
            TokenKind::Literal(Literal::Boolean(b)) => write!(f, "{}", b),
            TokenKind::Literal(Literal::Null) => write!(f, "Null"),
            TokenKind::Eof => write!(f, ""),
        }
    }
}

/// Contains the TokenType as well as the position.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Token {
    /// The kind of a token
    pub kind: TokenKind,
    /// Byte position
    pub position: (usize, usize),
}

impl Token {
    // TODO get rid of this
    pub fn identifier(&self) -> Result<String, InterpretError> {
        match self.kind() {
            TokenKind::Ident(ident) => Ok(ident.0.clone()),
            kind => Err(InterpretError::wrong_kind(kind)),
        }
    }

    // TODO get rid of this
    pub fn literal(&self) -> Result<&Literal, InterpretError> {
        match self.kind() {
            TokenKind::Literal(lit) => Ok(lit),
            kind => Err(InterpretError::wrong_kind(kind)),
        }
    }

    // TODO get rid of this
    pub fn sentinel() -> Token {
        Self {
            kind: TokenKind::X,
            position: (0, 0),
        }
    }

    pub fn start(&self) -> usize {
        self.position.0
    }

    pub fn end(&self) -> usize {
        self.position.1
    }

    pub(crate) fn span(&self) -> Span {
        // TODO remove this and make span the stored type
        Span::new(CharIndex(self.start()), CharIndex(self.end()))
    }
}

impl Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "'{}'", self.kind,)
    }
}

impl Token {
    pub fn kind(&self) -> &TokenKind {
        &self.kind
    }
}
