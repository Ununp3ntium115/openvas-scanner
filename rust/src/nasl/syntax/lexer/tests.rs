use crate::nasl::{
    syntax::{grammar::Declaration, Statement},
    Code,
};

pub fn parse_ok(file_name: &str, code: &str) -> Vec<Declaration> {
    let results = Code::from_string_fake_filename(code, file_name).parse();
    results.emit_errors().unwrap().decls()
}

pub fn parse_err(file_name: &str, code: &str) -> String {
    let results = Code::from_string_fake_filename(code, file_name).parse();
    results.unwrap_errors_str()
}

macro_rules! parse_test_ok {
    ($name: ident, $($code: literal$(,)?)*) => {
        #[test]
        fn $name() {
            $(
                insta::assert_snapshot!(crate::nasl::syntax::lexer::tests::parse_ok(
                    stringify!($name),
                        $code
                )
                .into_iter()
                .map(|stmt| stmt.to_string())
                .collect::<Vec<_>>()
                .join("\n"));
            )*
        }
    };
}

macro_rules! parse_test_err {
    ($name: ident, $($code: literal$(,)?)*) => {
        #[test]
        fn $name() {
            $(
                insta::assert_snapshot!(crate::nasl::syntax::lexer::tests::parse_err(
                    stringify!($name),
                        $code
                ));
            )*
        }
    };
}

pub(crate) use {parse_test_err, parse_test_ok};

use crate::nasl::syntax::token;

use super::*;

use StatementKind::*;
use TokenKind::*;

// simplified resolve method to verify a calculate with a given statement
fn resolve(s: &Statement) -> i64 {
    let callable = |stmts: &[Statement], calculus: Box<dyn Fn(i64, i64) -> i64>| -> i64 {
        let right = &stmts[1];
        let left = &stmts[0];
        calculus(resolve(left), resolve(right))
    };
    let single_callable = |stmts: &[Statement], calculus: Box<dyn Fn(i64) -> i64>| -> i64 {
        let left = &stmts[0];
        calculus(resolve(left))
    };
    match s.kind() {
        Primitive => match s.start().kind() {
            Literal(token::Literal::Number(num)) => *num,
            _ => todo!(),
        },
        Operator(head, rest) => match head {
            Tilde => single_callable(rest, Box::new(|left| !left)),
            Plus => callable(rest, Box::new(|left, right| left + right)),
            Minus if rest.len() == 1 => single_callable(rest, Box::new(|left| -left)),
            Minus => callable(rest, Box::new(|left, right| left - right)),
            Star => callable(rest, Box::new(|left, right| left * right)),
            Slash => callable(rest, Box::new(|left, right| left / right)),
            Percent => callable(rest, Box::new(|left, right| left % right)),
            LessLess => callable(rest, Box::new(|left, right| left << right)),
            GreaterGreater => callable(rest, Box::new(|left, right| left >> right)),
            Ampersand => callable(rest, Box::new(|left, right| left & right)),
            Pipe => callable(rest, Box::new(|left, right| left | right)),
            Caret => callable(rest, Box::new(|left, right| left ^ right)),
            GreaterGreaterGreater => {
                callable(
                    rest,
                    Box::new(|left, right| {
                        // this operator is used to drop signed bits
                        // so the result depends heavily if it is u32, u64, ...
                        // to have the same results as in javascript we use u32 in this example
                        let left_casted = left as u32;
                        (left_casted >> right) as i64
                    }),
                )
            }
            StarStar => callable(
                rest,
                Box::new(|left, right| (left as u32).pow(right as u32) as i64),
            ),
            token => {
                todo!("{:?}", token)
            }
        },
        _ => todo!("operator not found"),
    }
}

macro_rules! calculated_test {
    ($code:expr, $expected:expr) => {
        let expr = match parse_ok("", $code).remove(0) {
            Declaration::Statement(stmt) => stmt,
        };
        assert_eq!(resolve(&expr), $expected);
    };
}

#[test]
fn ordering() {
    calculated_test!("1 + 5 * 6;", 31);
    calculated_test!("3 * 10 + 10 / 5;", 32);
    calculated_test!("3 * 10 / 5;", 6);
    calculated_test!("3 * 10 / 5 % 4;", 2);
    calculated_test!("1 - 1 - 1;", -1);
    calculated_test!("1 - 1 * 2;", -1);
}

#[test]
fn grouping() {
    calculated_test!("(2 + 5) * 2;", 14);
}

#[test]
fn pow() {
    calculated_test!("2 ** 4;", 16);
}

#[test]
fn bitwise_operations() {
    //shifting
    calculated_test!("1 << 2 * 3;", 64);
    calculated_test!("3 * 12 >> 2;", 9);
    calculated_test!("-5 >>> 2;", 1073741822);
    // operations
    calculated_test!("1 & 0;", 0);
    calculated_test!("~1 | 0;", -2);
    calculated_test!("1 ^ 1;", 0);
}

parse_test_ok!(
    operator_assignment,
    "
    a += 1;
    a -= 1;
    a /= 1;
    a *= 1;
    a %= 1;
    a >>= 1;
    a <<= 1;
    a >>>= 1;
    "
);

parse_test_ok!(
    compare_operator,
    "
    a !~ '1';
    a =~ '1';
    a >< '1';
    a >!< '1';
    a == '1';
    a != '1';
    a > '1';
    a < '1';
    a >= '1';
    a <= '1';
    x() x 2;
    "
);

parse_test_ok!(logical_operator, "a && 1; a || 1;");

parse_test_ok!(assignment, "(a = 1);");

parse_test_ok!(variable_assignment_operator, "a++; a--; a[1]++; a[1]--;");

parse_test_ok!(primitive, "1;");
parse_test_ok!(variable, "a;");
parse_test_ok!(array, "a[1];");
parse_test_ok!(call, "a();");
parse_test_ok!(exit, "exit(0);");
parse_test_ok!(return_stmt, "return 0;");
parse_test_ok!(break_stmt, "break;");
parse_test_ok!(continue_stmt, "continue;");
parse_test_ok!(include, "include(\"test.inc\");");
parse_test_ok!(declare, "local_var a;");
parse_test_ok!(parameter, "[a, b];");
parse_test_ok!(named_parameter, "a: b;");
parse_test_ok!(assign, "a = 1;");
parse_test_ok!(add, "a + 1;");
parse_test_ok!(sub, "a - 1;");
parse_test_ok!(mul, "a * 1;");
parse_test_ok!(div, "a / 1;");
parse_test_ok!(modulo, "a % 1;");
parse_test_ok!(return_assign, "a++;");
parse_test_ok!(assign_return, "--a;");
parse_test_ok!(if_stmt, "if (a) b; else c;");
parse_test_ok!(for_stmt, "for (i = 0; i < 10; i++) a;");
parse_test_ok!(while_stmt, "while (a) b;");
parse_test_ok!(repeat, "repeat a; until b;");
parse_test_ok!(foreach, "foreach a(b) c;");
parse_test_ok!(block, "{ a; }");
parse_test_ok!(function_declaration, "function a(b) {c;}");
parse_test_ok!(no_op, ";");

parse_test_ok!(variables, "a;");
parse_test_ok!(arrays, "a[0];", "a = [1, 2, 3];", "a[0] = [1, 2, 4];");
parse_test_ok!(anon_function_call, "a(1, 2, 3);");
parse_test_ok!(
    named_function_call,
    "script_tag(name:\"cvss_base\", value:1 + 1 % 2);"
);
parse_test_ok!(use_parser, "a = 23;b = 1;");

// test_err!(wrong_assignment, "a = ");
// test_err!(wrong_keyword_assignment, "a = for;");

parse_test_ok!(
    position,
    r#"
        a = 1 + 1;
        b = 2 * 2;
        a = ++a;
        arr = mkarray(a, b, c      );
        arr[++a];
        exit(1);
        return 1;
        include('test.inc');
        local_var a, b, c;
        global_var a, b, c;
        if (a) display(1); else display(2);
        for (i = 1; i < 10; i++) display(i);
        while(TRUE) display(i);
        foreach a(q) display(a);
        repeat display("q"); until 1;
        {
            a;
            b;
        }
        function register_packages( buf ) { return 1; }
    "#
);

parse_test_ok!(
    change_to_peek,
    "send_packet( udp, pcap_active:FALSE ) x 200;"
);

parse_test_ok!(
    stack_overflow,
    r###"
        req = raw_string(0x00, 0x00, 0x03, 0x14, 0x08, 0x14, 0xff, 0x9f,
                0xde, 0x5d, 0x5f, 0xb3, 0x07, 0x8f, 0x49, 0xa7,
                0x79, 0x6a, 0x03, 0x3d, 0xaf, 0x55, 0x00, 0x00,
                0x00, 0x7e, 0x64, 0x69, 0x66, 0x66, 0x69, 0x65,
                0x2d, 0x68, 0x65, 0x6c, 0x6c, 0x6d, 0x61, 0x6e,
                0x2d, 0x67, 0x72, 0x6f, 0x75, 0x70, 0x2d, 0x65,
                0x78, 0x63, 0x68, 0x61, 0x6e, 0x67, 0x65, 0x2d,
                0x73, 0x68, 0x61, 0x32, 0x35, 0x36, 0x2c, 0x64,
                0x69, 0x66, 0x66, 0x69, 0x65, 0x2d, 0x68, 0x65,
                0x6c, 0x6c, 0x6d, 0x61, 0x6e, 0x2d, 0x67, 0x72,
                0x6f, 0x75, 0x70, 0x2d, 0x65, 0x78, 0x63, 0x68,
                0x61, 0x6e, 0x67, 0x65, 0x2d, 0x73, 0x68, 0x61,
                0x31, 0x2c, 0x64, 0x69, 0x66, 0x66, 0x69, 0x65,
                0x2d, 0x68, 0x65, 0x6c, 0x6c, 0x6d, 0x61, 0x6e,
                0x2d, 0x67, 0x72, 0x6f, 0x75, 0x70, 0x31, 0x34,
                0x2d, 0x73, 0x68, 0x61, 0x31, 0x2c, 0x64, 0x69,
                0x66, 0x66, 0x69, 0x65, 0x2d, 0x68, 0x65, 0x6c,
                0x6c, 0x6d, 0x61, 0x6e, 0x2d, 0x67, 0x72, 0x6f,
                0x75, 0x70, 0x31, 0x2d, 0x73, 0x68, 0x61, 0x31,
                0x00, 0x00, 0x00, 0x0f, 0x73, 0x73, 0x68, 0x2d,
                0x72, 0x73, 0x61, 0x2c, 0x73, 0x73, 0x68, 0x2d,
                0x64, 0x73, 0x73, 0x00, 0x00, 0x00, 0x9d, 0x61,
                0x65, 0x73, 0x31, 0x32, 0x38, 0x2d, 0x63, 0x62,
                0x63, 0x2c, 0x33, 0x64, 0x65, 0x73, 0x2d, 0x63,
                0x62, 0x63, 0x2c, 0x62, 0x6c, 0x6f, 0x77, 0x66,
                0x69, 0x73, 0x68, 0x2d, 0x63, 0x62, 0x63, 0x2c,
                0x63, 0x61, 0x73, 0x74, 0x31, 0x32, 0x38, 0x2d,
                0x63, 0x62, 0x63, 0x2c, 0x61, 0x72, 0x63, 0x66,
                0x6f, 0x75, 0x72, 0x31, 0x32, 0x38, 0x2c, 0x61,
                0x72, 0x63, 0x66, 0x6f, 0x75, 0x72, 0x32, 0x35,
                0x36, 0x2c, 0x61, 0x72, 0x63, 0x66, 0x6f, 0x75,
                0x72, 0x2c, 0x61, 0x65, 0x73, 0x31, 0x39, 0x32,
                0x2d, 0x63, 0x62, 0x63, 0x2c, 0x61, 0x65, 0x73,
                0x32, 0x35, 0x36, 0x2d, 0x63, 0x62, 0x63, 0x2c,
                0x72, 0x69, 0x6a, 0x6e, 0x64, 0x61, 0x65, 0x6c,
                0x2d, 0x63, 0x62, 0x63, 0x40, 0x6c, 0x79, 0x73,
                0x61, 0x74, 0x6f, 0x72, 0x2e, 0x6c, 0x69, 0x75,
                0x2e, 0x73, 0x65, 0x2c, 0x61, 0x65, 0x73, 0x31,
                0x32, 0x38, 0x2d, 0x63, 0x74, 0x72, 0x2c, 0x61,
                0x65, 0x73, 0x31, 0x39, 0x32, 0x2d, 0x63, 0x74,
                0x72, 0x2c, 0x61, 0x65, 0x73, 0x32, 0x35, 0x36,
                0x2d, 0x63, 0x74, 0x72, 0x00, 0x00, 0x00, 0x9d,
                0x61, 0x65, 0x73, 0x31, 0x32, 0x38, 0x2d, 0x63,
                0x62, 0x63, 0x2c, 0x33, 0x64, 0x65, 0x73, 0x2d,
                0x63, 0x62, 0x63, 0x2c, 0x62, 0x6c, 0x6f, 0x77,
                0x66, 0x69, 0x73, 0x68, 0x2d, 0x63, 0x62, 0x63,
                0x2c, 0x63, 0x61, 0x73, 0x74, 0x31, 0x32, 0x38,
                0x2d, 0x63, 0x62, 0x63, 0x2c, 0x61, 0x72, 0x63,
                0x66, 0x6f, 0x75, 0x72, 0x31, 0x32, 0x38, 0x2c,
                0x61, 0x72, 0x63, 0x66, 0x6f, 0x75, 0x72, 0x32,
                0x35, 0x36, 0x2c, 0x61, 0x72, 0x63, 0x66, 0x6f,
                0x75, 0x72, 0x2c, 0x61, 0x65, 0x73, 0x31, 0x39,
                0x32, 0x2d, 0x63, 0x62, 0x63, 0x2c, 0x61, 0x65,
                0x73, 0x32, 0x35, 0x36, 0x2d, 0x63, 0x62, 0x63,
                0x2c, 0x72, 0x69, 0x6a, 0x6e, 0x64, 0x61, 0x65,
                0x6c, 0x2d, 0x63, 0x62, 0x63, 0x40, 0x6c, 0x79,
                0x73, 0x61, 0x74, 0x6f, 0x72, 0x2e, 0x6c, 0x69,
                0x75, 0x2e, 0x73, 0x65, 0x2c, 0x61, 0x65, 0x73,
                0x31, 0x32, 0x38, 0x2d, 0x63, 0x74, 0x72, 0x2c,
                0x61, 0x65, 0x73, 0x31, 0x39, 0x32, 0x2d, 0x63,
                0x74, 0x72, 0x2c, 0x61, 0x65, 0x73, 0x32, 0x35,
                0x36, 0x2d, 0x63, 0x74, 0x72, 0x00, 0x00, 0x00,
                0x69, 0x68, 0x6d, 0x61, 0x63, 0x2d, 0x6d, 0x64,
                0x35, 0x2c, 0x68, 0x6d, 0x61, 0x63, 0x2d, 0x73,
                0x68, 0x61, 0x31, 0x2c, 0x75, 0x6d, 0x61, 0x63,
                0x2d, 0x36, 0x34, 0x40, 0x6f, 0x70, 0x65, 0x6e,
                0x73, 0x73, 0x68, 0x2e, 0x63, 0x6f, 0x6d, 0x2c,
                0x68, 0x6d, 0x61, 0x63, 0x2d, 0x72, 0x69, 0x70,
                0x65, 0x6d, 0x64, 0x31, 0x36, 0x30, 0x2c, 0x68,
                0x6d, 0x61, 0x63, 0x2d, 0x72, 0x69, 0x70, 0x65,
                0x6d, 0x64, 0x31, 0x36, 0x30, 0x40, 0x6f, 0x70,
                0x65, 0x6e, 0x73, 0x73, 0x68, 0x2e, 0x63, 0x6f,
                0x6d, 0x2c, 0x68, 0x6d, 0x61, 0x63, 0x2d, 0x73,
                0x68, 0x61, 0x31, 0x2d, 0x39, 0x36, 0x2c, 0x68,
                0x6d, 0x61, 0x63, 0x2d, 0x6d, 0x64, 0x35, 0x2d,
                0x39, 0x36, 0x00, 0x00, 0x00, 0x69, 0x68, 0x6d,
                0x61, 0x63, 0x2d, 0x6d, 0x64, 0x35, 0x2c, 0x68,
                0x6d, 0x61, 0x63, 0x2d, 0x73, 0x68, 0x61, 0x31,
                0x2c, 0x75, 0x6d, 0x61, 0x63, 0x2d, 0x36, 0x34,
                0x40, 0x6f, 0x70, 0x65, 0x6e, 0x73, 0x73, 0x68,
                0x2e, 0x63, 0x6f, 0x6d, 0x2c, 0x68, 0x6d, 0x61,
                0x63, 0x2d, 0x72, 0x69, 0x70, 0x65, 0x6d, 0x64,
                0x31, 0x36, 0x30, 0x2c, 0x68, 0x6d, 0x61, 0x63,
                0x2d, 0x72, 0x69, 0x70, 0x65, 0x6d, 0x64, 0x31,
                0x36, 0x30, 0x40, 0x6f, 0x70, 0x65, 0x6e, 0x73,
                0x73, 0x68, 0x2e, 0x63, 0x6f, 0x6d, 0x2c, 0x68,
                0x6d, 0x61, 0x63, 0x2d, 0x73, 0x68, 0x61, 0x31,
                0x2d, 0x39, 0x36, 0x2c, 0x68, 0x6d, 0x61, 0x63,
                0x2d, 0x6d, 0x64, 0x35, 0x2d, 0x39, 0x36, 0x00,

                ##3rd byte in this next line causes crash
                0x00, 0x00, 0x28, 0x7a, 0x6c, 0x69, 0x62, 0x40,
                0x6f, 0x70, 0x65, 0x6e, 0x73, 0x73, 0x68, 0x2e,
                0x63, 0x6f, 0x6d, 0x2c, 0x7a, 0x6c, 0x69, 0x62,
                0x2c, 0x6e, 0x6f, 0x6e, 0x65, 0x00, 0x00, 0x00,
                0x1a, 0x7a, 0x6c, 0x69, 0x62, 0x40, 0x6f, 0x70,
                0x65, 0x6e, 0x73, 0x73, 0x68, 0x2e, 0x63, 0x6f,
                0x6d, 0x2c, 0x7a, 0x6c, 0x69, 0x62, 0x2c, 0x6e,
                0x6f, 0x6e, 0x65, 0x00, 0x00, 0x00, 0x00, 0x00,
                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
                0x0a);
    "###
);

parse_test_ok!(
    jsp_example,
    r#"
    gms_path = gms_path + 'webapps\\appliance\\';
    jsp = '<% out.println( "' + jsp_print  + '" ); %>';
    "#
);

parse_test_ok!(
    unexpected_assign,
    r###"
    while(y = recv(socket:soc, length:1024)) {
    buf1 += y;
    }
    "###
);

parse_test_ok!(
    unexpected_noop,
    r###"
    if( ! version || version == '' ) return;
    "###
);

parse_test_ok!(
    unexpected_equal_operator,
    r###"
# Message Server runs on ports 36xx or 39xx
if (port < 3600 || port >= 3700)
if (port < 3900 || port >= 4000)
exit(0);

soc = open_sock_tcp(port);

    "###
);

parse_test_ok!(
    unexpected_plusplus,
    r###"
    cookie_jar[this_cookie]++;
    "###
);
