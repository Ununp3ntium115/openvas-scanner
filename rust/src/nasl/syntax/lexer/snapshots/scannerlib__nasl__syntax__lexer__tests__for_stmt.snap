---
source: src/nasl/syntax/lexer/tests.rs
expression: "crate :: nasl :: syntax :: lexer :: tests ::\nparse_ok(stringify! (for_stmt),\n\"for (i = 0; i < 10; i++) a;\").into_iter().map(| stmt |\nstmt.to_string()).collect :: < Vec < _ >> ().join(\"\\n\")"
snapshot_kind: text
---
for (i = 0; i < 10; i++) { a }
